pkg/proxy/service.go

```diff
@@ -110,36 +213,47 @@ func UpdateServiceMap(serviceMap ServiceMap, changes *ServiceChangeTracker) (res
 	// computing this incrementally similarly to serviceMap.
 	result.HCServiceNodePorts = make(map[types.NamespacedName]uint16)
 	for svcPortName, info := range serviceMap {
-		if info.HealthCheckNodePort() != 0 {
-			result.HCServiceNodePorts[svcPortName.NamespacedName] = uint16(info.HealthCheckNodePort())
+		if info.GetHealthCheckNodePort() != 0 {
+			result.HCServiceNodePorts[svcPortName.NamespacedName] = uint16(info.GetHealthCheckNodePort())
 		}
 	}
 
 	return result
 }
 
-// ServiceMap maps a service to its ServicePort information.
+// ServiceMap maps a service to its ServicePort.
 type ServiceMap map[ServicePortName]ServicePort
 
 // serviceToServiceMap translates a single Service object to a ServiceMap.
-// makeServicePort() return a proxy.ServicePort based on the given Service and its ServicePort.  We inject makeServicePort()
-// so that giving caller side a chance to initialize proxy.ServicePort interface.
 //
 // NOTE: service object should NOT be modified.
-func serviceToServiceMap(service *api.Service, makeServicePort func(servicePort *api.ServicePort, service *api.Service) ServicePort) ServiceMap {
+func (sct *ServiceChangeTracker) serviceToServiceMap(service *api.Service) ServiceMap {
 	if service == nil {
 		return nil
 	}
 	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
-	if proxyutil.ShouldSkipService(svcName, service) {
+	if utilproxy.ShouldSkipService(svcName, service) {
 		return nil
 	}
 
+	if len(service.Spec.ClusterIP) != 0 {
```

Comment:
Can we merge the ipv6 filtering logic into one place? 
Having ipv6 related logic scatter around will make it very hard to maintain.

Reply:
Is it possible to filter earlier? You just need the ServiceSpec right?

For instance, can we filter in `ServiceChangeTracker.Update`? And have a function to scan thru the service spec. If there is IP Version problem, then skip storing the service into ServiceMap?

Reply:
We can also introduce a state in ServiceMap to memorize that the ServiceSpec IP version is in an invalid state. As long as the ServiceSpec is still invalid, kube-proxy would not record event. This avoids events storm where each kube-proxy will generate an event on every update. 

Same applies for Endpoints. 

What do you think?

Reply:
Discussed with @freehan offline, we decided to go with this for now and opened https://github.com/kubernetes/kubernetes/issues/60471 for future improvement.

Reply:
Added more comments in code to make the filter purpose clearer.